crates/bevy_ecs/src/schedule/condition.rs

```diff
@@ -109,6 +156,31 @@ pub mod common_conditions {
         }
     }
 
+    /// Generates a [`Condition`](super::Condition)-satisfying closure that returns `true`
+    /// if the state machine changed state.
+    ///
+    /// To do things on transitions to/from specific states, use their respective OnEnter/OnExit
+    /// schedules. Use this run condition if you want to detect any change, regardless of the value.
+    ///
+    /// # Panics
+    ///
+    /// The condition will panic if the resource does not exist.
+    pub fn state_changed<S: States>() -> impl FnMut(Res<State<S>>) -> bool {
+        move |current_state: Res<State<S>>| current_state.is_changed()
+    }
+
+    /// Generates a [`Condition`](super::Condition)-satisfying closure that returns `true`
+    /// if there are any new events of the given type since it was last called.
+    pub fn on_event<T: Event>() -> impl FnMut(EventReader<T>) -> bool {
+        move |mut reader: EventReader<T>| reader.iter().count() > 0
```

Comment:
Should this be `!is_empty()`?

Reply:
That would be a bug. See https://github.com/IyesGames/iyes_loopless/issues/31

It used to be `!is_empty()` in iyes_loopless before, and we specifically changed it to `.iter().count()` after that bug was reported.

I guess, a more efficient way to fix it would be to clear the event reader after calling `is_empty()`, to make sure it does not retain pending events for the next time the run condition is executed.

Reply:
Got it, thanks. Either way, a note about the iterator-consumption being intentional wouldn't go amiss for me.

Reply:
Yeah, I'd like to see a comment in the code here.

Reply:
> I guess, a more efficient way to fix it would be to clear the event reader after calling `is_empty()`, to make sure it does not retain pending events for the next time the run condition is executed.

FYI this is now unnecessary due to #7582.

Reply:
Okay, leaving the iterator code.

Added an explanatory comment. It mentions both the correctness issue as well as the (lack of) performance consideration due to the iterator.